modules/js-and-css/speculation-rules/helper.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Helper functions used for Speculation Rules.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Prints the speculation rules in a cross-browser compatible way.
+ *
+ * For browsers that do not support speculation rules yet, the rules will not be loaded.
+ *
+ * @since n.e.x.t
+ *
+ * @return array Associative array of speculation rules by type.
+ */
+function plsr_get_speculation_rules() {
+	$prerender_rules = array(
+		array(
+			'source'    => 'document',
+			'where'     => array(
+				'and' => array(
+					// Prerender any URLs within the same site.
+					array(
+						'href_matches' => '/*\\\\?*',
+						'relative_to'  => 'document',
+					),
+					// Except for WP login and admin URLs.
+					array(
+						'not' => array(
+							'href_matches' => array(
+								'/wp-login.php\\\\?*#*',
+								'/wp-admin/*\\\\?*#*',
+							),
+						),
+					),
+					// And except for any links marked with a class to not prerender.
+					array(
+						'not' => array(
+							'selector_matches' => '.no-prerender',
+						),
+					),
+				),
+			),
+			'eagerness' => 'moderate',
+		),
+	);
+
+	return array( 'prerender' => $prerender_rules );
```

Comment:
I suppose eventually this should be filterable?

Reply:
+1 I agree we could use a more declarative API for both inclusion and exclusion lists, or maybe even a way of tagging elements with rules (plus filters of course).

This came up talking to optimization plugins that already implement prefetch/render either manually or with scripts like instant.page or quicklinks. These plugins need to make sure they exclude specific routes that might affect state, like clicking a signup link. They have to both discover and maintain these links, usually introduced by other plugins.

Ideally core would offer a way for other plugins to register routes to exclude from prefetch/render. Similarly we could offer a way to register routes you want to make sure to prefetch/render.

Reply:
I think for now the simple exclude paths filter allows the most critical integration, that should work for an MVP.

Eventually it's worth building out a more bespoke API, but the nesting of these rules can also be quite complex, and I would want to make sure we keep the complexity contained so that external developers have to worry about it as little as possible.

It's also worth thinking further about what _should_ be possible to be modified vs what would be discouraged from the core side.

I'd say let's go with the simple version and explore a more advanced API in a future issue, some time after this has shipped.